x-pack/plugins/rule_registry/server/rule_data_plugin_service/resource_installer.ts

```diff
@@ -0,0 +1,416 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { get, isEmpty, once } from 'lodash';
+import { estypes } from '@elastic/elasticsearch';
+
+import { ElasticsearchClient, Logger } from 'kibana/server';
+
+import {
+  DEFAULT_ILM_POLICY_ID,
+  ECS_COMPONENT_TEMPLATE_NAME,
+  TECHNICAL_COMPONENT_TEMPLATE_NAME,
+} from '../../common/assets';
+import { technicalComponentTemplate } from '../../common/assets/component_templates/technical_component_template';
+import { ecsComponentTemplate } from '../../common/assets/component_templates/ecs_component_template';
+import { defaultLifecyclePolicy } from '../../common/assets/lifecycle_policies/default_lifecycle_policy';
+
+import { ResourceNames } from './resource_names';
+import { IndexNames } from './index_names';
+import { IndexOptions } from './index_options';
+import { incrementIndexName } from './utils';
+
+export enum Resources {
+  common = 'common resources shared between all indices',
+  forIndex = 'resources for a particular index',
+  forNamespace = 'resources for a particular namespace',
+}
+
+const INSTALLATION_TIMEOUT = 60000;
+
+interface ConstructorOptions {
+  resourceNames: ResourceNames;
+  getClusterClient: () => Promise<ElasticsearchClient>;
+  logger: Logger;
+  isWriteEnabled: boolean;
+}
+
+export class ResourceInstaller {
+  constructor(private readonly options: ConstructorOptions) {}
+
+  private getResourceName(...relativeNameSegments: string[]) {
+    return this.options.resourceNames.getFullName(...relativeNameSegments);
+  }
+
+  public memoizeInstallation<T>(
+    resources: Resources,
+    installResources: () => Promise<T>
+  ): () => Promise<T> {
+    return once(async () => {
+      try {
+        return await Promise.race([
+          installResources(),
+          new Promise<T>((resolve, reject) => {
+            setTimeout(() => {
+              const msg = `Timeout: it took more than ${INSTALLATION_TIMEOUT}ms`;
+              reject(new Error(msg));
+            }, INSTALLATION_TIMEOUT);
```

Comment:
This timeout logic was there before, and I didn't want to remove it. Moreover, I added some error handling on top of it to make sure:

- exceptions are logged asap and with meaningful error messages
- exceptions are propagated up to the callers, so that:
  - failed installation of common resources will cause failure in the installation of index resources
  - failed installation of index resources will cause failure in methods of `RuleDataWriter` and `RuleDataReader`

We didn't have this at all and I think it's important to have this kind of error handling.

As for the timeout logic, it's a good question. I see both pros and cons. If ES cluster or network is overloaded, we will have much longer request-response times during index bootstrapping. We execute quite a bunch of requests during bootstrapping (moreover this number depends on the number of indices), so the resulting time could end up being very high. All read/write operations will be blocked until the bootstrapping is complete. Which means:

- If we don't have a combined timeout, all the rules will be hanging, and there won't be any indication in the UI or logs.
- If we have a timeout, users will see failing rules in the UI, and there will be some Kibana logs to check. This sounds better to me, although not perfect.

I'd say let's keep timeouts for now, maybe increase the value to a couple of minutes. Good UX will require more thought on that (e.g. maybe we don't need to block read operations until the bootstrapping is complete; maybe we need some indication in the UI that bootstrapping is in progress/succeeded/failed, etc); we could raise this topic in Alerts As Data sync later.

Reply:
I left the timeout but changed it to 20 minutes.
These 20 minutes are applied to installation of common and index-level resources separately.